src/ILCompiler.Compiler/src/Compiler/MetadataGeneration.cs

```diff
@@ -133,6 +137,9 @@ private void Graph_NewMarkedNode(DependencyNodeCore<NodeFactory> obj)
                     return;
                 }
 
+                if (method.HasInstantiation && !method.IsAbstract)
+                    ExactMethodInstantiations.AddExactMethodInstantiationEntry(_nodeFactory, method);
```

Comment:
This node could have a separate phase that we trigger just before saving the native layout node.

It's not desirable for nodes to have a mutable state because it makes threading difficult. We haven't made a pass on multithreading yet, but I don't want us to end up with dozens of locks protecting various private summary HashSets when we can only have one hash set that is the source of truth for everything else. (We will probably even get rid of the HashSet in MetadataGeneration and query the NodeFactory directly, so there won't even be that one HashSet.)